man/authdb_manage.8
.\" Copyright Georgios Portokalidis, 2004.
.\"
.\" This file may be copied under the conditions described
.\" in the GNU GENERAL PUBLIC LICENSE, Version 2, June 1991
.\" that should have been distributed together with this file.
.\"
.TH AUTHDB_MANAGE 8 "May 18, 2004" "Version 0.8" "System Manager's Manual"
.SH NAME
authdb_manage \- Admission control resources database manager
.SH SYNOPSIS
.BI "authdb_manage [" R "] " db_directory " " db_filename
.SH DESCRIPTION
A console menu driven interface to manage the databases used by admission
control for resource control.
.br
.IR db_directory " specifies the directory where the database files are located
and
.I db_filename
is the name of the file containing the databases. The optional parameter
.I R
can be used to try and recover a corrupted database, in case for example
authd(8) crashes. The databases need to be in Berkeley DB (http:/
www.sleepycat.com) format and should be stored in a single file.
.P
.B authdb_manage
can be used to manage:
.IP \(bu 2
Function names
.IP \(bu 2
Library names
.IP \(bu 2
Available resources for library\-function pairs
.IP \(bu 2
Resource consumption of library\-function pairs
.P
When first run,
.B authdb_manage
prints a menu with the above mentioned entities.
When one of them is selected the user is presented with the following database
actions:
.P
.BR "1. List" "   [ list stored entities ]"
.br
.BR "2. Insert" " [ insert a new entity ]"
.br
.BR "3. Delete" " [ delete an entity ]"
.br
.BR "0. Return" " [ return to the previous menu ]"
.P
Below we describe the different entities that can be manipulated.
.P
.B FUNCTIONS
.br
Each function must have a unique name and a 32 bits integer key. Only the lower
16 bits of the key can be set by the user and it shouldn't be 0. Note that
these values are not checked for correctness by
.BR "authdb_manage" .
.br
Additionally a function can be set to have 'global resource consumption' which
means that its resource consumption is the same for all libraries. This will
cause the upper 16 bits of the key to be asserted.
.P
.B LIBRARIES
.br
A library must have a unique name and a 32 bits integer key. Only the lower 16
bits of the key can be set by the user and it shouldn't be 0. Note that these
values are not checked for correctness by
.BR "authdb_manage" .
.P
.B RESOURCES
.br
A resource must have a unique 32 bits integer key and it shouldn't be zero. Note
that this value is not checked for correctness by
.B "authdb_manage" .
.br
Resource availability is also represented by a 32 bits integer. Optionally a
description string can also be specified.
.P
.B RESOURCE CONSUMPTION
.br
Resource consumption is specified for each library\-function pair.
The user is prompted to select a library and a function from the ones existing
in the databases (when a function is set to have 'global resource consumption',
the user is still prompted for a library but the selection is ignored).
.br
Resource consumption has a fixed part and a variable one. The fixed one is a 32
bit integer value representing the fixed cost to be assigned to each occurrence
of the function by resource control. The variable cost is calculated at runtime
by taking into account the values of the function's arguments. For this reason
the user is prompted to enter a formula that will be used to calculate this
cost. The formula is a simple arithmetic expression that can contain the
following operators: +, -, *, /, (, ). The values of the function's arguments
can be inserted in the formula by using printf like notation. For example an
integer 2nd argument would be represented as '%2$d' and 1st unsigned long long
argument as '%1$llu'. In case of strings their size is replaced in the formula
and functions are ignored.
.SH EXAMPLES
.B "authdb_manage /etc/authd/resctrl resource.db"
.P
Start the manager for the database file \'resource.db\' located in \'/etc/authd
resctrl\'
.SH SEE ALSO
authd(8)
.SH AUTHOR
Georgios Portokalidis (gportoka@liacs.nl)
.SH NOTES
If
.B authdb_manage
is not terminated properly the latest changes to the databases
might not have been stored. Even tough Ctrl-C from the keyboard is blocked,
users should take special care to exit the application normally.
.br
This utility is only provided as quick way to setup a resource control database
and is not in any was the best way to manage a continuously growing database.
.br
Special care should be take when setting variable cost formulas, since bad
illegal argument types and numbers, could cause authd(8) to crash.


ChangeLog
ADMCTRL - Admission Control Daemon - ChangeLog
--------------------------------------------

0.8.9
=====

Configuration
  * configure.ac Fixed for additional compatibility
  * src/Makefile.am Updated to reflect changes in configure.ac
  * tests/Makefile.am Updated to reflect changes in configure.ac
  * kernel/kadmctrl_req.c Removed the 'k' prefix from functions
  * kernel/kclient.c Updated to reflect changes in kernel/kadmctrl_req.c

0.8.8
=====

Configuration
  * configure.ac Added options to specify seperate keynote header and 
  library directory,

Tests
  * tests/enc_nonce Encrypts a nonce and prints it out in hex.
  * kernel/kclient.c Kernel driver test client.

Kernel Driver
  * kernel/kadmctrl_req.c Admission control request manipulation function
  for the kernel.

0.8.7
=====

Resource control
  * arith_parser.c updated postfix_expr_parse() & infix_expr_parse() to
  successfully parse an empty string and return 0 as a result.

0.8.6
=====

Documentation
  Added authd_setup(8) man page.
  Updated authd(3) man page.

Admission control client library
  Function admctrl_req_set_authinfo() now accepts buffers
  containing the public key and credentials instead of the files containing 
  them.
  Removed function admctrl_req_encrypt_nonce(). It's an end-user assisting
  function. Code provided as a C snippet in 'doc/encrypt_nonce.c'.

0.8.5
====

Documentation
  Added more documentation under doc/.
  Written tests/README.

Configuration
  Corrected configure.ac, src/Makefile.am and tests/Makefile.am to link only
  necessary libraries.

Tests
  Updated client to support resource control and test function type arguments.

0.8.4
=====

Documentation
  Added authd(3) man page for developers and updated documentation in doc/.

Configuration
  Added option --disable-authd to enable stand alone compilation of the client
  library.

0.8.3 
=====

Admission control front-end
  Added support for linux kernel module device driver.
  Renamed front-end executable to 'authdfe'.

Configuration
  Changed --without-clientlib to --disable-clientlib.
  Changed --with-frontend to --enable-front-end.
  Changed --with-keynote-prefix to --with-keynote.
  Changed --with-db-prefix to --with-berkeley-db.
  Changed --with-pthread-prefix to --with-pthread.
  Changed --with-snprintfv-prefix to --with-snprintfv.

0.8.2
=====

Added a linux kernel module device driver

0.8
===

Started using the GNU build system


src/authdb_manage.c
/* authdb_manage.c

  Copyright 2004  Georgios Portokalidis <digital_bull@users.sourceforge.net>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>

/** \file authdb_manage.c
	\brief Simple resource control db management application
  \author Georgios Portokalidis
	*/

#include "resource_ctrl.h"

static char *db_dir,*db_fn;
static resource_ctrl_db_t db;
static u_int32_t recover;

static void
process_args(int argc,char **argv)
{
	if ( argc < 3 || (argc > 3 && *argv[1] != 'R') )
	{
		fprintf(stderr,"%s: Illegal arguments\n",argv[0]);
		fprintf(stderr,"Syntax: %s [R] (db directory) (db filename)\n",argv[0]);
		exit(1);
	}

	if ( argc == 3 )
	{
		db_dir = argv[1];
		db_fn = argv[2];
	}
	else
	{
		recover = DB_RECOVER;
		db_dir = argv[2];
		db_fn = argv[3];
	}
}

static int
get_selection(const char *message,int min,int max)
{
	int selection,c;

	// Get a valid selection
	do {
		printf("%s",message);
		selection = getchar() - '0';
	} while( selection < min || selection > max);

	// Ignore until EOF or newline
	while ( (c = getchar()) != EOF && c != '\n' )
		;
	return selection;
}

static uint32_t
get_uint32(const char *message)
{
	char buf[64];
	int c,i;

	printf("%s",message);

	i = 0;
	while( (c = getchar()) != EOF && c != '\n' )
		if ( i < 63 )
			buf[i++] = (char)c;
	buf[i] = '\0';
	return strtoul(buf,NULL,10);
}

static void
get_string(const char *message,size_t max,char *buf)
{
	int c;
	unsigned int i;

	printf("%s",message);

	i = 0;
	--max;
	while( (c = getchar()) != EOF && c != '\n' )
		if ( i < max )
			buf[i++] = (char)c;
	buf[i] = '\0';
}

static int
confirm(const char *message,int sel1)
{
	int c;
	int ret = -1;

	printf("%s",message);
	if ( getchar() == sel1 )
		ret = 0;

	// Ignore until EOF or newline
	while ( (c = getchar()) != EOF && c != '\n' )
		;

	return ret;
}

static void
print_menu()
{
	printf("MAIN MENU\n");
	printf("1. Functions\n");
	printf("2. Libraries\n");
	printf("3. Resources\n");
	printf("4. Resource consumption\n");
	printf("0. Exit\n");
}

static void
print_menu2()
{
	printf(" ACTIONS\n");
	printf("1. List\n");
	printf("2. Insert\n");
	printf("3. Delete\n");
	printf("0. Return\n");
}


static void
functions()
{
	int sel;
	u_int32_t uint;
	char buf[MAX_ACTION_NAME_SIZE];

	do {
		printf("FUNCTIONS");
		print_menu2();
		printf("Selection:");
		sel = get_selection("Selection:",0,3);
		putchar('\n');
		switch( sel )
		{
			case 0:
				break;
			case 1:
				printf("Function list:\n");
				resource_ctrl_display_functions(&db);
				break;
			case 2:
				printf("Adding new function...\nExisting functions:\n");
				resource_ctrl_display_functions(&db);
				get_string("Insert new function name:",MAX_ACTION_NAME_SIZE,buf);
				uint = get_uint32("Insert new function key:");
        if ( confirm("Is this function's resource consumption global(Y/N)? ",'Y') == 0 )
          uint |= (~0U) << 16;
        else
          uint &= (~0U) >> 16;
				if ( resource_ctrl_add_function(&db,buf,uint) == 0 )
					printf("Function added!\n");
				else
					printf("Insertion failed!\n");
				break;
			case 3:
				printf("Deleting function...\nExisting functions:\n");
				resource_ctrl_display_functions(&db);
				get_string("Enter function to delete:",MAX_ACTION_NAME_SIZE,buf);
				if ( confirm("Are you sure(Y/N)?",'Y') != 0 )
					break;
				if ( resource_ctrl_del_function(&db,buf) == 0 )
					printf("Function and associated resource consumption has been deleted!\n");
				else
					printf("Deletion failed!\n");
				break;
			default:
				printf("Invalid selection!\n");
		}
		putchar('\n');
	} while (sel != 0);
}

static void
libraries()
{
	int sel;
	u_int32_t uint;
	char buf[MAX_ACTION_NAME_SIZE];

	do {
		printf("LIBRARIES");
		print_menu2();
		sel = get_selection("Selection:",0,3);
		putchar('\n');
		switch( sel )
		{
			case 0:
				break;
			case 1:
				printf("Libraries list:\n");
				resource_ctrl_display_libraries(&db);
				break;
			case 2:
				printf("Adding new library:\nExisting libraries:\n");
				resource_ctrl_display_libraries(&db);
				get_string("Insert new library name:",MAX_ACTION_NAME_SIZE,buf);
				uint = get_uint32("Insert new library key:");
        uint &= (~0) >> 16;
				if ( resource_ctrl_add_library(&db,buf,uint) == 0 )
					printf("Library added!\n");
				else
					printf("Insertion failed!\n");
				break;
			case 3:
				printf("Deleting library...\nExisting libraries:\n");
				resource_ctrl_display_libraries(&db);
				get_string("Enter library to delete:",MAX_ACTION_NAME_SIZE,buf);
				fgets(buf,MAX_ACTION_NAME_SIZE,stdin);
				if ( confirm("Are you sure(Y/N)?",'Y') != 0 )
					break;
				if ( resource_ctrl_del_library(&db,buf) == 0 )
					printf("Library and associated resource consumption has been deleted!\n");
				else
					printf("Deletion failed!\n");
				break;
			default:
				printf("Invalid selection!\n");
		}
		putchar('\n');
	} while (sel != 0);
}

static void
consumption()
{
	int sel;
	u_int32_t pk,rkey;
	char func[MAX_ACTION_NAME_SIZE],lib[MAX_ACTION_NAME_SIZE];
	resource_consumption_t consumption;

	do {
		printf("RESOURCE CONSUMPTION");
		print_menu2();
		sel = get_selection("Selection:",0,3);
		if ( sel <= 0 )
			break;
		printf("Existing functions:\n");
		resource_ctrl_display_functions(&db);
		get_string("Select function:",MAX_ACTION_NAME_SIZE,func);
		printf("Existing libraries:\n");
		resource_ctrl_display_libraries(&db);
		get_string("Select library:",MAX_ACTION_NAME_SIZE,lib);
		if ( resource_ctrl_resourcekey(&db,func,lib,&pk) != 0 )
		{
			printf("Error acquiring key for supplied combination!\n\n");
			continue;
		}
		switch( sel )
		{
			case 1:
				printf("Resource consumption list:\n");
				resource_ctrl_display_consumption(&db,pk);
				break;
			case 2:
				printf("Adding resource consumption...\nExisting resources:\n");
				resource_ctrl_display_resources(&db);
				consumption.rkey = get_uint32("Select resource key:");
				consumption.fixed_cost = get_uint32("Enter new resource consumption fixed cost:");
				get_string("Enter new resource consumption variable cost formula:",RESOURCE_CTRL_MAX_VAR_FORM_LEN,consumption.variable_cost_formula);
				if ( resource_ctrl_add_consumption(&db,pk,&consumption) == 0 )
					printf("Resource consumption added!\n");
				else
					printf("Insertion failed!\n");
				break;
			case 3:
				printf("Deleting resource consumption..\nExisting:\n");
				resource_ctrl_display_consumption(&db,pk);
				rkey = get_uint32("Enter resource key to delete:");
				if ( confirm("Are you sure(Y/N)?",'Y') != 0 )
					break;
				if ( resource_ctrl_del_consumption(&db,pk,rkey) == 0 )
					printf("Resource consumption has been deleted!\n");
				else
					printf("Deletion failed!\n");
				break;
			default:
				printf("Invalid selection!\n");
		}
		putchar('\n');
	} while (sel != 0);
}

static void
resources()
{
	int sel;
	u_int32_t uint;
	resource_t resource;

	do {
		printf("RESOURCES");
		print_menu2();
		sel = get_selection("Selection:",0,3);
		putchar('\n');
		switch( sel )
		{
			case 0:
				break;
			case 1:
				printf("Resources list:\n");
				resource_ctrl_display_resources(&db);
				break;
			case 2:
				printf("Adding new resource...\nExisting resources:\n");
				resource_ctrl_display_resources(&db);
				uint = get_uint32("Insert new resource key:");
				resource.available = get_uint32("Insert new resource availability:");
				get_string("Insert new resource description:",RESOURCE_CTRL_MAX_RES_DESCR_LEN,resource.description);
				if ( resource_ctrl_add_resource(&db,uint,&resource) == 0 )
					printf("Resource added!\n");
				else
					printf("Insertion failed!\n");
				break;
			case 3:
				printf("Deleting resource...\nExisting resources:\n");
				resource_ctrl_display_resources(&db);
				uint = get_uint32("Enter resource key to delete:");
				if ( confirm("Are you sure(Y/N)?",'Y') != 0 )
					break;
				if ( resource_ctrl_del_resource(&db,uint) == 0 )
					printf("All resource records have been removed!\n");
				else
					printf("Deletion failed!\n");
				break;
			default:
				printf("Invalid selection!\n");
				break;
		}
		putchar('\n');
	} while (sel != 0);
}

static void
process_selection(int sel)
{
	switch( sel )
	{
		case 0:
			break;
		case 1:
			functions();
			break;
		case 2:
			libraries();
			break;
		case 3:
			resources();
			break;
		case 4:
			consumption();
			break;
		default:
			break;
	}
}


int 
main(int argc,char **argv)
{
	int e,selection;

	process_args(argc,argv);

	if ( (e = resource_ctrl_dbinit(&db)) != 0 )
	{
		fprintf(stderr,"%s: Error initialising DB\n",argv[0]);
		return 1;
	}

	if ( (e = resource_ctrl_dbopen(&db,db_dir,db_fn,DB_CREATE,DB_CREATE | recover)) != 0 )
	{
		fprintf(stderr,"%s: Error opening DB\n",argv[0]);
		return 1;
	}

	do {
		print_menu();
		selection = get_selection("Selection:",0,4);
		process_selection(selection);
	} while( selection != 0 );

	resource_ctrl_dbclose(&db);
	
	return 0;
}
